### PR TITLE
Improve exceptions

### PR DESCRIPTION
**Pull request by gumb0**
Fixes https://github.com/ethereum/cpp-ethereum/issues/4845

Having a `std::string` member in an exception class goes against boost recommendations (see http://www.boost.org/doc/libs/1_39_0/libs/exception/doc/exception_types_as_simple_semantic_tags.html and http://www.boost.org/community/error_handling.html)
and leads to weird behaviour.
(`boost::enable_current_exception()` copies the exception object provided to `BOOST_THROW_EXCEPTION` macro, but this string member is not copied; `boost::enable_current_exception()` actually requires copy constructor to never throw)

Also there's no point in storing additional error string as a `std::string` member because `error_info` mechanism of boost exceptions can and should be used to store any additional info about the error.

First commit here was changing `Exception` constructor to store the string in `error_info` instead of a member variable.
Then I decided to get rid of all exception constructors taking a parameter string and use the approach more natural to boost execption with using `operator<<` at the point of throw.

### CODE REVIEW

**Comment by chfast** [comment]
I finally realized  what I don't like about the Boost approach -- it's too "dynamic".

1. Throwing an exception is quite complex exception now, quite a lot of boilerplate.
2. It's hard to know what an exception contains. I cannot jump to the exception declaration and inspect it fields. At runtime in can contain multiple error_infos but it's hard to tell if any and where they come from.

If we need to add additional data on the fly, can we do it like this?

```
catch(Exception& ex)
{
    ex.addComment("comment");
    throw;
}
```

**Comment by gumb0** [comment]
What you suggest is even more dynamic, you couldn't then easily check whether the data of specific type is present in `Exception`. And what if I want to analyze an integer error code in `catch` ?

**Comment by gumb0** [comment]
Boilerplate - on the one hand yes, on the other hand something like
```
        BOOST_THROW_EXCEPTION(FailedToOpenDB() << errinfo_db(status.ToString()) << errinfo_path(_path));
```
is less bolierplate comparing to declaring special exception class with error code and path...

**Comment by gumb0** [comment]
In general this dynamic approach seems unusual to me, too

Once I worked with a more static approach to this, something like
```
class Exception
{
public:
  Exception() {}
  Exception(Exception const& _suberror) {}
};
...
catch(LowerLevelException& ex)
{
  throw HigherLevelException(ex);
}
```
`LowerLevelException` and `HigherLevelException` could have some additional data members specific to them.
And if you throw with a macro you could even save the file position of each throw along the stack.

**Comment by gumb0** [comment]
Anyway in this PR I'm fixing the bug in the current approach.

If it's too ugly, look at the first commit, it's very simple, it keeps `std::string` as a constructor argument and it works, too.

**Comment by gumb0** [comment]
> Maybe some error messages will become worse after this change, because outputting ex.what() will show only the exception name now (it's better to output boost::diagnostic_information(ex) in this case)

Fixed this by returning `boost::diagnostic_information` from `what`

File: libdevcore/Exceptions.h [hunk 1]
```diff
@@ -21,34 +21,38 @@
...
-	Exception(std::string _message = std::string()): m_message(std::move(_message)) {}
-	const char* what() const noexcept override { return m_message.empty() ? std::exception::what() : m_message.c_str(); }
-
-private:
-	std::string m_message;
+    const char* what() const noexcept override { return boost::diagnostic_information_what(*this); }
 };
 
-#define DEV_SIMPLE_EXCEPTION(X) struct X: virtual Exception { const char* what() const noexcept override { return #X; } }
+#define DEV_SIMPLE_EXCEPTION(X)  \
```
Comment (gumb0) [inline libdevcore/Exceptions.h]: We could live without these macros now, actually

**Comment by gumb0** [comment]
Boost actually advertises this dynamic nature as the strength of boost::exception
> Deriving from boost::exception effectively decouples the semantics of a failure from the information that is relevant to each individual instance of reporting a failure with a given semantic.
>
> In other words: with boost::exception, what data a given exception object transports depends primarily on the context in which failures are reported (not on its type.)

**Pull request merged**